### DISCUSSION TITLE
[PS 8.1.1] Smaller service locator in admin controller

### DISCUSSION THREAD

**Discussion by d0niek** [PrestaShop/PrestaShop#33805] [Q&A]
Hi!

I've upgraded my PS `v1.7.8.10` to `v.8.1.1` via `autoupgrade` module (cli) and I'm really confusing. In my [modern admin controllers](https://devdocs.prestashop-project.org/8/modules/concepts/controllers/admin-controllers/) I'm using other services via `$this->get('service_name')` and I got an error:

```
Service "service_name" not found: even though it exists in the app's container, the container inside "MyVendor\Controller\Admin\MyNameController" is a smaller service locator that only knows about the "doctrine", "form.factory", "http_kernel", "parameter_bag", "request_stack", "router", "security.authorization_checker", "security.csrf.token_manager", "security.token_storage", "serializer", "session" and "twig" services. Try using dependency injection instead.
```

It's Ok, I have to use **DI** and all works! But ... is it works in "yours" controllers? For example [here](https://github.com/PrestaShop/PrestaShop/blob/8.1.x/src/PrestaShopBundle/Controller/Admin/FrameworkBundleAdminController.php#L212) you're calling service which doesn't exists. How? Why?

My question is, my upgrade broke something with container or you forgot to update calls? o0

Of course in `v1.7.8.10` everything was work.

**Comment by kpodemski** [comment]
Could you paste the service.yml of your module? Did you make your module compatible with Symfony 4.4?

Make sure to follow these guidelines:
https://devdocs.prestashop-project.org/8/modules/core-updates/8.0/#symfony-update

especially: Services that are not dependency injected and that are not declared as “public” are removed from the container

**Reply by d0niek** [reply]
Nice url/tip! Thanks!

With my services it's Ok, they will be dependency injected ;-) My working (in PS `v1.7.8.8+`) **service.yml** file looks like:

```yml
services:
  _defaults:
    autoconfigure: true
    autowire: true
    public: true

  MyVendor\Controller\Admin\MyNameController:
    class: MyVendor\Controller\Admin\MyNameController

  # etc.
```

I gave wrong an example. The problem is with "yours" services. In my controller when I call `$this->getContext()` or `$this->trans(...)` I got an error I mention above. The service `prestashop.adapter.legacy.context` or `translator` is unavailable. They are not public?

The question is, my upgrade broke something or yours services aren't updated to Symfony `v4.4`?

**Comment by dr-woland** [comment]
No, you did not break anything.

PrestaShop's controllers used to inherit from a [controller class](https://github.com/symfony/framework-bundle/blob/4.4/Controller/Controller.php) from the Symfony's FrameworkBundle that had been deprecated in Symfony 4.2 and was removed in 5.0. This class implemented `ContainerAwareInterface` so the full app container was always injected during [controller resolution](https://github.com/symfony/symfony/blob/4.4/src/Symfony/Bundle/FrameworkBundle/Controller/ControllerResolver.php#L70) (this will be removed in Symfony 7.0).

Currently, controllers extend a [`AbstractController`](https://github.com/symfony/framework-bundle/blob/4.4/Controller/AbstractController.php), so they are [service subscribers](https://symfony.com/doc/4.4/service_container/service_subscribers_locators.html) - instead of the full app container they should be injected a container instance that includes only services specified in the `getSubscribedServices` method. For this to work, the controller service definition needs to have a `container.service_subscriber` tag - in your case, it comes from the autoconfiguration. If the tag was missing (and the container was not injected using any other way), the FrameworkBundle's `ControllerResolver` would raise a deprecation and inject the  `service_container` service.

The reason why core controllers in 8.x have access to the full container and do not generate above mentioned deprecation errors is because of the [container pass](https://github.com/PrestaShop/PrestaShop/blob/8.1.x/src/PrestaShopBundle/DependencyInjection/Compiler/ContainerInjectionPass.php) that adds `service_container` injection to controller service definitions marked with the `prestashop.core.controllers` tag. This behaviour will be [removed in 9.0](https://github.com/PrestaShop/PrestaShop/pull/32450).

**Reply by d0niek** [reply]
Hi @dr-woland!

Nice reply!

Ok, so should I refactor all my controllers to be full dependency injection (I have to inject `translator`, `context`, etc. separately)? It should work on Prestashop `v1.7.8.*` and after upgrade nothing won't breaks (and I'm ready to `v9.0` too) (?).
